Firestore/core/src/local/local_store.cc

```diff
@@ -161,22 +180,36 @@ LocalWriteResult LocalStore::WriteLocally(std::vector<Mutation>&& mutations) {
   }
 
   return persistence_->Run("Locally write mutations", [&] {
+    // Figure out which keys do not have a remote version in the cache, this is
+    // needed to create the right overlay mutation: if no remote version
+    // presents, we do not need to create overlays as patch mutations.
+    // TODO(Overlay): Is there a better way to determine this? Document version
+    // does not work because local mutations set them back to 0.
+    auto remote_docs = remote_document_cache_->GetAll(keys);
+    std::unordered_set<DocumentKey, DocumentKeyHash>
+        docs_without_remote_version;
+    for (const auto& entry : remote_docs) {
+      if (!entry.second.is_valid_document()) {
+        docs_without_remote_version.insert(entry.first);
+      }
+    }
     // Load and apply all existing mutations. This lets us compute the current
     // base state for all non-idempotent transforms before applying any
     // additional user-provided writes.
-    DocumentMap existing_documents = local_documents_->GetDocuments(keys);
+    auto overlayed_documents =
+        local_documents_->GetOverlayedDocuments(remote_docs);
 
     // For non-idempotent mutations (such as `FieldValue.increment()`), we
     // record the base state in a separate patch mutation. This is later used to
     // guarantee consistent values and prevents flicker even if the backend
     // sends us an update that already includes our transform.
     std::vector<Mutation> base_mutations;
     for (const Mutation& mutation : mutations) {
-      absl::optional<Document> base_document =
-          existing_documents.get(mutation.key());
-
+      auto it = overlayed_documents.find(mutation.key());
+      HARD_ASSERT(it != overlayed_documents.end(),
```

Comment:
Consider including the key value into the assertion message; this will help with debugging if this assertion were to ever fail.